x-pack/plugins/lens/public/indexpattern_plugin/dimension_panel/settings.tsx

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import React, { useState } from 'react';
+import { i18n } from '@kbn/i18n';
+import {
+  EuiPopover,
+  EuiButtonIcon,
+  EuiFlexItem,
+  EuiContextMenuItem,
+  EuiContextMenuPanel,
+} from '@elastic/eui';
+import { IndexPatternColumn } from '../indexpattern';
+import { IndexPatternDimensionPanelProps } from './dimension_panel';
+import {
+  getColumnOrder,
+  operationDefinitionMap,
+  getOperations,
+  getOperationDisplay,
+} from '../operations';
+
+export interface SettingsProps extends IndexPatternDimensionPanelProps {
+  selectedColumn: IndexPatternColumn;
+  filteredColumns: IndexPatternColumn[];
+}
+
+export function Settings({
+  selectedColumn,
+  filteredColumns,
+  state,
+  columnId,
+  setState,
+}: SettingsProps) {
+  const [isSettingsOpen, setSettingsOpen] = useState(false);
+  const contextOptionBuilder =
+    selectedColumn && operationDefinitionMap[selectedColumn.operationType].contextMenu;
+  const contextOptions = contextOptionBuilder
+    ? contextOptionBuilder({
+        state,
+        setState,
+        columnId,
+      })
+    : [];
+  const operations = getOperations();
+  const operationPanels = getOperationDisplay();
+  const functionsFromField = selectedColumn
+    ? filteredColumns.filter(col => {
+        return (
+          (!('sourceField' in selectedColumn) && !('sourceField' in col)) ||
+          ('sourceField' in selectedColumn &&
+            'sourceField' in col &&
+            col.sourceField === selectedColumn.sourceField)
+        );
+      })
+    : filteredColumns;
+
+  const operationMenuItems = operations
+    .filter(o => selectedColumn && functionsFromField.some(col => col.operationType === o))
+    .map(o => (
+      <EuiContextMenuItem
+        data-test-subj={`lns-indexPatternDimension-${o}`}
+        key={o}
+        icon={selectedColumn && selectedColumn.operationType === o ? 'check' : 'empty'}
+        onClick={() => {
+          const newColumn: IndexPatternColumn = filteredColumns.find(
+            col =>
+              col.operationType === o &&
+              (!('sourceField' in col) ||
+                !('sourceField' in selectedColumn) ||
+                col.sourceField === selectedColumn.sourceField)
+          )!;
```

Comment:
I moved it into a helper function, unfortunately because of how typeguards work i can only move the basic check for one column into a helper, not the compound checks of multiple columns.